### PR TITLE
Deferred deep linking

### PR DESCRIPTION
**Pull request by AGulev**
More info about this feature available here https://developers.facebook.com/docs/app-ads/deep-linking/

### CODE REVIEW

File: facebook/src/facebook.cpp [hunk 1]
```diff
@@ -57,6 +57,21 @@ static int Facebook_GetVersion(lua_State* L)
...
+    int callback = dmScript::Ref(L, LUA_REGISTRYINDEX);
+
+    dmScript::GetInstance(L);
+    int context = dmScript::Ref(L, LUA_REGISTRYINDEX);
+
+    lua_State* thread = dmScript::GetMainThread(L);
+
+    Platform_FetchDeferredAppLinkData(L, callback, context, thread);
+    
+    return 1;
```
Comment (mathiaswking) [inline facebook/src/facebook.cpp]: We usually check that we have pushed that number of return values.
E.g. use DM_LUA_STACKCHECK and similar mechanics

File: facebook/src/facebook_android.cpp [hunk 2]
```diff
@@ -553,6 +573,21 @@ int Facebook_ShowDialog(lua_State* L)
...
+    VerifyCallback(L);
+    g_Facebook.m_Callback_DeferredAppLink = callback;
+    g_Facebook.m_Self_DeferredAppLink = context;
+
+    JNIEnv* environment = Attach();
+    environment->CallVoidMethod(g_Facebook.m_FB, g_Facebook.m_FetchDeferredAppLink, (jlong)thread);
+
+    if (!Detach(environment))
+    {
+        dmLogError("An unexpected error occurred during Facebook JNI interaction.");
```
Comment (mathiaswking) [inline facebook/src/facebook_android.cpp]: Why not return a lua error? (I see other functions that also have this issue)

E.g. "return luaL_error(L, "err msg");"

File: facebook/src/facebook_android.cpp [hunk 3]
```diff
@@ -553,6 +573,21 @@ int Facebook_ShowDialog(lua_State* L)
...
+    }
+}
```
Comment (mathiaswking) [inline facebook/src/facebook_android.cpp]: Other functions here check the Lua stack

File: facebook/src/facebook_emscripten.cpp [hunk 4]
```diff
@@ -361,6 +361,13 @@ int Facebook_DisableEventUsage(lua_State* L)
     return 0;
 }
 
+int Facebook_FetchDeferredAppLinkData(lua_State* L)
+{
+    dmLogDebug("get_deferred_app_link() function doesn't support on HTML5 platform");
```
Comment (mathiaswking) [inline facebook/src/facebook_emscripten.cpp]: "doesn't support" -> "isn't supported"

Comment (mathiaswking) [inline facebook/src/facebook_emscripten.cpp]: dmLogDebugOnce

File: facebook/src/java/com/defold/facebook/FacebookJNI.java [hunk 5]
```diff
@@ -287,4 +291,33 @@ public void disableEventUsage() {
         facebook.enableEventUsage();
     }
 
+    public void fetchDeferredAppLinkData(final long userData) {
+    AppLinkData.fetchDeferredAppLinkData(this.activity,
```
Comment (mathiaswking) [inline facebook/src/java/com/defold/facebook/FacebookJNI.java]: Seems like some indentation issue?

File: facebook/src/facebook_android.cpp (hunk 3 above)
Comment (AGulev) [reply facebook/src/facebook_android.cpp]: Yes! Thanks!

File: facebook/src/facebook.cpp (hunk 1 above)
Comment (AGulev) [reply facebook/src/facebook.cpp]: Nice catch, thx!

File: facebook/src/facebook_android.cpp (hunk 2 above)
Comment (AGulev) [reply facebook/src/facebook_android.cpp]: Fixed!

File: facebook/src/facebook_emscripten.cpp (hunk 4 above)
Comment (AGulev) [reply facebook/src/facebook_emscripten.cpp]: Thx! ☺️

Comment (AGulev) [reply facebook/src/facebook_emscripten.cpp]: `dmLogOnceDebug` - Done!

File: facebook/src/java/com/defold/facebook/FacebookJNI.java (hunk 5 above)
Comment (AGulev) [reply facebook/src/java/com/defold/facebook/FacebookJNI.java]: Fixed!

**Pull request merged**